subgraphs/osmosis/src/modules/Swap.ts

```diff
@@ -0,0 +1,353 @@
+  data: cosmos.TransactionData
+): void {
+  const events = data.tx.result.events;
+  for (let idx = 0; idx < events.length; idx++) {
+    if (events[idx].eventType == "token_swapped") {
+      swapEventHandler(events[idx], data.tx, data.block);
+    }
+  }
+}
+
+function swapEventHandler(
+  data: cosmos.Event,
+  tx: cosmos.TxResult,
+  block: cosmos.HeaderOnlyBlock
+): void {
+  const attributes = data.attributes;
+  let poolId: string = "";
+  let tokenInAmount: BigInt = constants.BIGINT_ZERO;
+  let tokenInDenom: string = "";
+  let tokenOutAmount: BigInt = constants.BIGINT_ZERO;
+  let tokenOutDenom: string = "";
+
+  for (let i = 0; i < attributes.length; i++) {
+    const key = attributes[i].key;
+    const value = attributes[i].value;
+    if (key == "pool_id") {
+      poolId = value;
+    } else if (key == "tokens_in" || key == "tokens_out") {
+      const j = tokenDataParser(value) as i32;
+      const tokenAmount = BigInt.fromString(value.substring(0, j));
+      const tokenDenom = value.substring(j, value.length);
+      if (key == "tokens_in") {
+        tokenInAmount = tokenAmount;
+        tokenInDenom = tokenDenom;
+      } else {
+        tokenOutAmount = tokenAmount;
+        tokenOutDenom = tokenDenom;
+      }
+    }
+  }
+
+  const liquidityPoolId = constants.Protocol.NAME.concat("-").concat(poolId);
+  const liquidityPool = LiquidityPoolStore.load(liquidityPoolId);
+  if (!liquidityPool) {
+    return;
+  }
+
+  swap(
+    poolId,
+    tokenInAmount,
+    tokenInDenom,
+    tokenOutAmount,
+    tokenOutDenom,
+    data,
+    tx,
+    block
+  );
+}
+
+function tokenDataParser(tokenData: string): number {
+  let i = 0;
+  while (i < tokenData.length) {
+    const tokeChar = tokenData.charAt(i);
+    if (
+      tokeChar == "0" ||
+      tokeChar == "1" ||
+      tokeChar == "2" ||
+      tokeChar == "3" ||
+      tokeChar == "4" ||
+      tokeChar == "5" ||
+      tokeChar == "6" ||
+      tokeChar == "7" ||
+      tokeChar == "8" ||
+      tokeChar == "9"
+    ) {
+      i++;
+    } else {
+      break;
+    }
+  }
+
+  return i;
+}
+
+function swap(
+  poolId: string,
+  tokenInAmount: BigInt,
+  tokenInDenom: string,
+  tokenOutAmount: BigInt,
+  tokenOutDenom: string,
+  data: cosmos.Event,
+  tx: cosmos.TxResult,
+  block: cosmos.HeaderOnlyBlock
+): void {
+  const liquidityPoolId = constants.Protocol.NAME.concat("-").concat(poolId);
+  const liquidityPool = LiquidityPoolStore.load(liquidityPoolId);
+  if (!liquidityPool) {
+    return;
+  }
+
+  const inputTokenBalances = liquidityPool.inputTokenBalances;
+  const inputTokenAmounts = new Array<BigInt>(inputTokenBalances.length).fill(
+    constants.BIGINT_ZERO
+  );
+  const inputTokens = liquidityPool.inputTokens;
+  const tokenInIndex = inputTokens.indexOf(tokenInDenom);
+  if (tokenInIndex >= 0) {
+    inputTokenBalances[tokenInIndex] = inputTokenBalances[tokenInIndex].plus(
+      tokenInAmount
+    );
+    inputTokenAmounts[tokenInIndex] = tokenInAmount;
```

Comment:
As it is a plus (+) operation and tokenInAmount should > 0, in theory the balance after plus should be > 0. Add a log.error() here in case there is something wrong at 21703cbe62252f89cb45c85d27afd200b45fc0ca